R/methods-epi_archive.R

```diff
@@ -59,9 +59,10 @@
 #'
 #' @importFrom data.table between key
 #' @export
-epix_as_of <- function(x, version, min_time_value = -Inf, all_versions = FALSE,
+epix_as_of <- function(x, version = NULL, min_time_value = -Inf, all_versions = FALSE,
                        max_version = deprecated()) {
   assert_class(x, "epi_archive")
+  version <- version %||% min(max(x$DT$time_value), x$versions_end)
```

Comment:
Isn't this going to chop off the latest version available?  `max(x$DT$time_value)` is `< x$versions_end` for most surveillance data.  Could we do just `x$versions_end`?
```suggestion
  version <- version %||% x$versions_end
```
Also, due to that vctrs bug, I think if `min` stays, it needs to be something like `min(c(<a>, <b>))` rather than `min(<a>, <b>)` for now.  (Probably an issue for yearmonth for example.)

(Also, thoughts on introducing a new function for this rather than a default?)

Reply:
`x$versions_end` sounds reasonable there, I didn't think much about the time_value bit, just copying a pattern from elsewhere. 

A separate function would make sense if this default is unexpected or there are multiple competing reasonable defaults. Is this such a case? Some default possibilities
- versions_end
- Sys.Date()
Others? The first one seems more reasonable to me as a default.

Reply:
I think the first option (versions_end) is the simplest approach, and makes sense for "get the latest".  The second one... seems like "get now", but raises various complications and will be harder to implement, but ultimately could be more useful.  This is probably rehashing [this](https://github.com/cmu-delphi/epiprocess/issues/109), but here's some foreseen issues:
- Sys.Date() vs. time_type disagreements
- Querying beyond `versions_end`: likely a very common complaint, and we should at least have a `with/set_versions_end` helper or something like that.  Way back in the day I was considering having some automatic leniency here, but it seems potentially dependent on the time_type what the leniency should be, requiring thought, so I set this idea aside.  (At the time, I was thinking that some of this type of logic & clobberable versions logic might ultimately go in a new "updating epi archive" class/classes, and might happen earlier than epix_as_of calls.)

So... not sure here whether we should go with the convenient thing or maybe the more useful thing which will catch when the "latest" actually isn't anytime recent because we forgot/failed to update the archive with fresh data.

I feel it could be confusing to have "get the latest" or "get now" be the default for epix_as_of(); for the uninitiated this could seem obtuse:
```r
my_archive %>%
  epix_as_of() %>%
  arx_forecaster(......)
```
Trying to evaluate this as if I had never seen epiprocess before: if I were to read that, I'd expect that should be an error ("as of what?"), but I guess that'd also force me to read the relevant docs and understand, but probably still wouldn't be happy.

Reply:
When I started using archives, I distinctly recall being puzzled, and a bit frustrated, with epix_as_of(Sys.Date()) breaking every time. And then trying epix_as_of(.$max_version) and then epix_as_of(.$max_versions) and then giving up and looking at the documentation. I never clarified why max(version) < versions_end was not allowed, but I accepted it, figuring there were reasons.

I agree that from a readability POV that basic example doesn't look great and this reads better

```r
my_archive %>%
  epix_latest() %>% # or maybe epix_as_of_latest()
  arx_forecaster(......)
```

I think I'm convinced that epix_latest()/epix_as_of_latest() can justify breaking the "there should be one way to do it" rule, for better readability. Could try implementing it in this PR. I see a couple possibilities:

```r
# We continue to disallow max(version) < versions_end
epix_latest <- function(arch) {
  # inspect arch time_type and construct the current_time_value
  arch %>% epix_as_of(min(c(current_time_value, .$versions_end)))
}
# Allow max(versions) < versions_end
epix_latest <- function(arch) {
  # inspect arch time_type and construct the current_time_value
  arch %>% epix_as_of(current_time_value) 
}
# ... probably more work in epi_archive to automatically do min(max(versions), versions_end) where it matters
```